pkg/services/dashboards/database/database.go

```diff
@@ -994,16 +997,220 @@ func (d *dashboardStore) FindDashboards(ctx context.Context, query *dashboards.F
+UNION ALL
+SELECT org_id, 'folder' as kind, uid, id, folder_uid as parent_uid, title FROM dashboard WHERE title {{ .Dialect.LikeStr }} {{ .Arg .Text }} AND is_folder
+{{ end }}
+
+SELECT DISTINCT
+	entity.org_id, entity.kind, entity.uid, entity.id, entity.title,
+	dashboard_tag.term as term, f1.uid as folder_uid, f1.title as folder_title FROM (
+	{{- template "text-search-legacy" . -}}
+) AS entity
+LEFT JOIN folder f1 ON (entity.parent_uid = f1.uid AND entity.org_id = f1.org_id)
+{{ if ne .UserID 0 }}   {{- /* If a user is not admin: need to filter by permissions */ -}}
+LEFT JOIN folder f2 ON (f1.parent_uid     = f2.uid AND f1.org_id     = f2.org_id)
+LEFT JOIN folder f3 ON (f2.parent_uid     = f3.uid AND f2.org_id     = f3.org_id)
+LEFT JOIN folder f4 ON (f3.parent_uid     = f4.uid AND f3.org_id     = f4.org_id)
+INNER JOIN permission p ON 
+p.role_id IN (
+	SELECT ur.role_id FROM user_role AS ur    WHERE ur.user_id = {{ .Arg .UserID }} AND (ur.org_id = {{ .Arg .OrgID }} OR ur.org_id = 0)
+	UNION
+	SELECT tr.role_id FROM team_role AS tr    WHERE tr.team_id IN (
+		{{- range $i, $el := .TeamIDs -}}
+		{{- if $i -}},{{- end -}}
+		{{- $.Arg $el -}}
+		{{- end -}}
+	) AND tr.org_id = {{ .Arg .OrgID }}
+	UNION
+	SELECT br.role_id FROM builtin_role AS br WHERE br.role IN ({{ .Arg .Role }}) AND (br.org_id = {{ .Arg .OrgID }} OR br.org_id = 0)
+) AND (
+{{ if .HasSplitScopes }}
+	{{- /* This permission check relies on the optimised (indexed) split scope columns in the permission table */ -}}
+	(
+		p.kind = 'dashboards' AND
+		p.action = 'dashboards:read' AND
+		p.identifier = entity.uid
+	) OR (
+		p.kind = 'folders' AND
+		p.action = 'folders:read' AND
+		p.identifier IN (f4.uid, f3.uid, f2.uid, f1.uid, entity.uid)
+	)
+{{ else }}
+	(
+		p.action = 'dashboards:read' AND
+		p.scope LIKE 'dashboards:uid:%' AND
+		substr(p.scope, 16) = entity.uid
+	) OR (
+		p.action = 'folders:read' AND
+		p.scope LIKE 'folders:uid:%' AND
+		substr(p.scope, 13) IN (f4.uid, f3.uid, f2.uid, f1.uid, entity.uid)
+	)
+{{ end }}
+)
+{{ end }}
+LEFT JOIN dashboard_tag ON dashboard_tag.dashboard_id = entity.id
+ORDER BY entity.title ASC
+LIMIT {{ .Arg .Limit }}
+`))
+
+func (d *dashboardStore) findDashboards(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
+	// Only handle non-empty search queries
+	if query.Title == "" {
+		return nil, nil
+	}
+
+	// Get OrgID, Team IDs and UserID (zero for admins)
+	orgID, userID, teamIDs := query.OrgId, int64(0), []int64{}
+	if orgID == 0 {
+		orgID = query.SignedInUser.GetOrgID()
+	}
+	if !query.SignedInUser.GetIsGrafanaAdmin() {
+		namespaceID, identifier := query.SignedInUser.GetNamespacedID()
+		switch namespaceID {
+		case identity.NamespaceUser, identity.NamespaceServiceAccount:
+			userID, _ = identity.IntIdentifier(namespaceID, identifier)
+		}
+		teamIDs = query.SignedInUser.GetTeams()
+
+		// TODO: check wildcard dashboard and folder scopes?
```

Comment:
I think we should aim to do this if possible. It significantly sped the old query up. But I guess it might not have such a big impact with the new query if it's way faster in general.

Reply:
This is related to Kalle's remark and I moved it into a separate issue in the related [search epic](https://github.com/grafana/grafana-backend-platform-squad/issues/38) as it requires more work/tests. For now I consider it to be a known bug.